scanner/src/main/kotlin/ScannerCriteria.kt

```diff
@@ -0,0 +1,107 @@
+ */
+typealias ScannerConfigMatcher = (String) -> Boolean
+
+/**
+ * A data class defining selection criteria for scanners.
+ *
+ * An instance of this class is passed to a [ScanResultsStorage] to define the criteria a scan result must match,
+ * so that it can be used as a replacement for a result produced by an actual scanner. A scanner implementation
+ * creates a criteria object with its exact properties. Users can override some or all of these properties to
+ * state the criteria under which results from a storage are acceptable even if they deviate from the exact
+ * properties of the scanner. That way it can be configured for instance, that results produced by an older
+ * version of the scanner can be used.
+ */
+data class ScannerCriteria(
+    /**
+     * Criterion to match the scanner name. This can be an exact scanner name, but this property is actually a
+     * regular expression. So it is possible select multiple scanners using an alternative ('|') expression or
+     * an arbitrary one using a wildcard ('.*').
+     */
+    val regScannerName: String,
+
+    /**
+     * Criterion to match for the minimum scanner version. Results are accepted if they are produced from scanners
+     * with at least this version.
+     */
+    val minScannerVersion: Semver,
+
+    /**
+     * Criterion to match for the maximum scanner version. Results are accepted if they are produced from scanners
+     * with a version lower than this one. (This bound of the version range is excluding.)
+     */
+    val maxScannerVersion: Semver,
+
+    /**
+     * A function to check whether the configuration of a scanner is compatible with this criteria object.
+     */
+    val configMatcher: ScannerConfigMatcher
+) {
+    companion object {
+        /**
+         * A matcher for scanner configurations that accepts all configurations passed in. This function can be
+         * used if the concrete configuration of a scanner is irrelevant.
+         */
+        val ALL_CONFIG_MATCHER: ScannerConfigMatcher = { _ -> true }
+
+        /**
+         * A matcher for scanner configurations that accepts only exact matches of the [originalConfig]. This
+         * function can be used by scanners that are extremely sensitive about their configuration.
+         */
+        fun exactConfigMatcher(originalConfig: String): ScannerConfigMatcher = { config -> originalConfig == config }
+    }
+
+    /** The regular expression pattern to match for the scanner name. */
+    private val namePattern: Pattern by lazy { Pattern.compile(regScannerName) }
+
+    /**
+     * Check whether the [details] specified match the criteria stored in this object. Return true if and only if
+     * the result described by the [details] fulfill all the requirements expressed by the properties of this
+     * object.
+     */
+    fun isCompatible(details: ScannerDetails): Boolean {
+        if (!namePattern.matcher(details.name).matches()) {
```

Comment:
It's maybe a bit nicer to use the `Regex` class instead of `Pattern`, then you could do `details.name.matches(nameRegex)` here (or the other way round).